SputnikVM Integration

sorpaas opened the pull request:
@splix @sudachen @whilei Would love to get some primary review as I'm not too familiar with Go.

So how this works is that we directly inject multivm into the state processor. A switch if statement (currently `if true` so multivm never gets executed right now) detects whether to continue processing the state using go-ethereum's vm or a multivm implementation. Then it processes the vm results, and apply it to statedb.

A multivm implementor (a.k.a SputnikVM) would need to implement two interfaces. `MultiVmFactory` shared across the whole client, initialized once. `Create` function accepts a header and a transaction, and give out a vm. The `MultiVm` interface then implements the actual execution.

sudachen commented:
https://github.com/sudachen/go-ethereum-classic/blob/4ac038f00b5301ca04d4df84c9491900a7832fe8/core/vmenv.go

I don't understand what you want to do. It's all already implemented, as I see.

sorpaas commented:
@sudachen We have the December SputnikVM integration milestone to catch up so unfortunately I'll have to re-do/re-port/duplicate some of the work and get it ready for code review. I don't think this conflicts with your VM performance analysis project, however.

sudachen commented:
Wei, are you kidding me?! I already implemented all required to work with new VM, implemented old VM in this maner, and tested it. But not pulled only because it was not required in the moment.  I am finishing it with binary integration, but you come and say - let's do it another way! And starting coding wthout understanding what is done.  Hm, it's amazing!  By the way, geth is not your personal project as sputnikvm.

sorpaas commented:
@sudachen No offense, but @ia and I asked you many times to give an update on your work, or at least give an explanation of what you're planning to do. You haven't replied to any of the messages.

https://github.com/ethereumproject/go-ethereum/pull/349#issuecomment-338255797
https://github.com/ethereumproject/go-ethereum/pull/349#issuecomment-341436519

We also have a group chat on Slack which you joined to discuss the integration, but you've never replied to both @ia's and my messages.

All those happened at least two weeks ago. I value your work, but we should also respect the deadline.

sudachen commented:
There is really done. We need only binary integration with rust.

sudachen commented:
Although, why am I worried. Do this with "if", with two branches of tests, with two branch of working way ... to the more interesting support -) Why not.  I think you understand that you can test you code only with sputnik VM? So you need to test Integration and sputnik VM in one pice? And you can't use this integration with another VM like classicVM. With respect to deadline ... Are yo already know how you will bind rust sputnik lib with geth? Because I bind it with existing integration code. 

Related to plans, by the way, on summit I have showed to you integration code. Did not not?

sorpaas commented:
Thanks to `sputnikvm-ffi`, this branch is working right now and I'm able to sync to at least a large part of the network using the SputnikVM processor. So I'm planning to use it to conduct another regression test for SputnikVM using this branch (whose time can be reduced to several hours / one or two day). Help appreciated for code review and/or testing. `go test ./...` also mostly passes now, except for several ones which the processor cannot figure out the correct Patch to use.

From my reading of the code, it looks like `go-ethereum` uses heap allocated big integers and hash representations, which could be a bottleneck for performance currently. So I have a feeling that SputnikVM might be able to be faster right now, although we need to test it.

Usage of this branch:

```
go get github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm
cd $GOPATH/github.com/ethereumproject/sputnikvm-ffi/c
make build
```

And then build geth with `CGO_LDFLAGS`:

```
CGO_LDFLAGS=$GOPATH/github.com/ethereumproject/sputnikvm-ffi/c/libsputnikvm.a -ldl go build ./...
```

Run geth with the `--sputnikvm` flag to enable the SputnikVM processor:

```
./geth --sputnikvm --verbosity 6
```

The fastest way to test is to export your current Ethereum blockchain, use a new folder to re-import it.

```
./geth export blockchain.db
mv ~/.ethereum-classic ~/.ethereum-classic-old
./geth import blockchain.db
```

sorpaas commented:
I'm on block 1295000 without any failure. Everything looks green right now. The time it takes is around 10000 transactions per 30 seconds. So would be glad to have some review.

@splix @whilei @tzdybal

sudachen commented:
I'm totally disagree with this solution. It's not match to existing tests, adding additional way to execute contracts, cannot be measured, tested, proofed in equal way with existing execution path. It's very dirty. Also, in this solution VM is doing things which it not to do. 

By the way, i think, this solution will not work with static linking for some reasons.

sorpaas commented:
Well @sudachen, have you actually tried it...

* The test would pass if you just pass the `CGO_LDFLAGS` environment variable. That's why we'll need a Makefile or as @whilei mentioned, a Megafile.
* The test would also execute using the SputnikVM processor if you set UseSputnikVM to true (which can also be done in command line at build time).

`CGO_LDFLAGS` are also used in other well-known go libraries, like gorocksdb (https://github.com/tecbot/gorocksdb). We use static linking already here so there's no problem as you said... with static linking. What's more, I really don't know why you call a documented and tested-to-be-passing-the-regression-test clean code would be called "dirty" in your mouth comparing with 9000+ LOC changes without even any commit messages...

Since you have tried to attack me multiple times for reasons I don't even want to tell... I'll be honest with you starting from this point. In the past two months, you've never tried to collaborate with me and/or work with any other team members to share your design/idea. In the past two weeks, all what you've done is to introduce new not-possible-to-review code and never addressed concerns from other team members.

I actually finished this SputnikVM integration totally by accident. I was playing around with `sputnikvm-ffi` and figured out the integration is much easier to get done than I thought. We've had a discussion already about SputnikVM integration in August, and as the author of SputnikVM, I must say that **SputnikVM is designed to integrate with Geth in this simple way described in this PR**. If you ever listened once to what others have said, you would have avoided your 9000+ LOCs of disaster.

If you think VM should only do a subset of the thing described here, you should argue with the author of the Yellow Paper, Gavin Wood, but not me...

sorpaas commented:
The regression passed 2.5 million blocks without an issue. 🎏 

From this point on, I don't have the blockchain data locally so it needs to use [sputnikvm1](https://seashells.io/v/9Pn7ANmH). Would be glad if someone can help test. @whilei @tzdybal

sorpaas commented:
(Also just a side note. We currently have 6 patches (including ECIP1017) in ETC chain. Only four of them are related to EVM. They're Frontier, Homestead, EIP150 and EIP160. That's why you see in the state processor it only checks those four hard forks.)

sudachen commented:
Ok. Simple is not right. However, it's not important.  As, I say before I disagree with this solution, but I will not block this changes  as you do typically, because have have respect for the work of other people.  

By the way, standart tests will not work with your solution, as well a all code wich use core.NewEnv.

I can trim my code to only sputnik VM integration, and it will has small count of changes.  But with continuose implementation in existing architecture. Can you do that?

sorpaas commented:
My Internet connection seems fast today. Just passed 4 million blocks in localhost. Everything still looks green.

1 more million to go...

sorpaas commented:
Reached 4.9 million blocks in my localhost. Regression test is now finished and no errors reported! 🎏

whilei commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,163 @@
+			break Loop
+		case sputnikvm.RequireAccount:
+			address := ret.Address()
+			if statedb.Exist(address) {
```
In Go there's a general convention to avoid `else` statements, preferring "breaking" patterns instead, so we could use
```go
		case sputnikvm.RequireAccountCode:
			address := ret.Address()
			if statedb.Exist(address) {
				vm.CommitAccountCode(address, statedb.GetCode(address))
                break
			}
			vm.CommitNonexist(address)
```
noting that `break` will break the switch, not the `for` `Loop`.

whilei replied:
It's only a style thing, though.

sorpaas replied:
Hmm, I'll fix this. But can you explain why? Code above is actually harder to read for me compared with an else statement.

sorpaas replied:
Fixed.

tzdybal replied:
I totally disagree with @whilei. This is perfectly legal use of `else` statement. They are typically omitted only when code itself returns/continues/breaks, etc., see [Effective Go](https://golang.org/doc/effective_go.html#if).
I think that `if/else` block is much more readable.

whilei replied:
Noted @tzdybal -- I definitely don't have my heart set on it, and think your take is valid; in fact I think it's more readable as Wei wrote it originally too. Whatever you want, @sorpaas

whilei commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,163 @@
+	receipt.TxHash = tx.Hash()
+	receipt.GasUsed = new(big.Int).Set(totalUsedGas)
+	if MessageCreatesContract(tx) {
+		from, _ := tx.From()
```
Don't we already have `from` from up top on line 23?

sorpaas replied:
Does golang has variable scope? So this line defines a new `from` variable, and out of this `if` scope `from` would still be the above one, right? It is really common to do this kind of thing in Rust, does it work in go?

Anyway, `from` is never used below again so this shouldn't cause any trouble.

whilei replied:
I feel like I might be not understanding something here

sorpaas replied:
Ooops you're right. Fixed by removing this line.

whilei commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,163 @@
+
+	glog.V(logger.Debug).Infoln(receipt)
+
+	vm.Free()
```
Would it save any performance to bump `vm.Free()` up to after where it is last used in the function (line 144)? I like the "readability" of having it at the bottom, but just a thought.

sorpaas replied:
I don't think so. It would just mean that VM is freed early.

The problem is that Go only execute one transaction at a time, so it's always one VM created, one VM freed. We don't save any time by moving it up.

whilei replied:
roger that

whilei replied:
and we can't just slap the whole thing into a go routine because there might be conflicts in state, yes?

sorpaas replied:
Won't be any conflict. Each VM executes independently and they never share any memory. But if indeed we try to execute multiple VMs in parallel, moving it up might give us a really small performance improvement.

However, right now I prefer readability. Besides I don't think this is the performance bottleneck. We can always modify it easily in a later PR.

whilei commented on core/state_processor.go:
```diff
@@ -85,12 +85,21 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB) (ty
-		receipt, logs, _, err := ApplyTransaction(p.config, p.bc, gp, statedb, header, tx, totalUsedGas)
-		if err != nil {
-			return nil, nil, totalUsedGas, err
+		if UseSputnikVM == false {
```
Again here only a style thing

```go
	for i, tx := range block.Transactions() {
		if tx.Protected() {
			chainId := p.config.GetChainID()
			if chainId.Cmp(new(big.Int)) == 0 {
				return nil, nil, nil, fmt.Errorf("ChainID is not set for EIP-155 in chain configuration at block number: %v. \n  Tx ChainID: %v", block.Number(), tx.ChainId())
			}
			if tx.ChainId() == nil || tx.ChainId().Cmp(chainId) != 0 {
				return nil, nil, nil, fmt.Errorf("Invalid transaction chain id. Current chain id: %v tx chain id: %v", p.config.GetChainID(), tx.ChainId())
			}
		}
		statedb.StartRecord(tx.Hash(), block.Hash(), i)
		if !UseSputnikVM {
			receipt, logs, _, err := ApplyTransaction(p.config, p.bc, gp, statedb, header, tx, totalUsedGas)
			if err != nil {
				return nil, nil, totalUsedGas, err
			}
			receipts = append(receipts, receipt)
			allLogs = append(allLogs, logs...)
			continue
		}
		receipt, logs, _, err := ApplyMultiVmTransaction(p.config, p.bc, gp, statedb, header, tx, totalUsedGas)
		if err != nil {
			return nil, nil, totalUsedGas, err
		}
		receipts = append(receipts, receipt)
		allLogs = append(allLogs, logs...)
	}
```

whilei commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,163 @@
+Loop:
+	for {
+		ret := vm.Fire()
+		switch ret.Typ() {
```
Just double checking here... would it be safer/faster to use `ret.(type)` to use go's type switch instead of an exported function?

sorpaas replied:
`ret` is always of `Require` type. `ret.Typ()` returns an "enum" that contains the actual type (which in Go if I understand correctly, is an integer).

sorpaas replied:
Here's the GoDoc for Require if that's confusing: https://godoc.org/github.com/ethereumproject/sputnikvm-ffi/go/sputnikvm#Require

tzdybal commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,163 @@
+	UseSputnikVM = false
+)
+
+func ApplyMultiVmTransaction(config *ChainConfig, bc *BlockChain, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, totalUsedGas *big.Int) (*types.Receipt, evm.Logs, *big.Int, error) {
```
I would extract some methods to make this function shorter, but honestly its quite straight forward, so it's not a necessity.

whilei commented on core/multivm_processor.go:
```diff
@@ -0,0 +1,160 @@
+
+var UseSputnikVM = false
+
+func ApplyMultiVmTransaction(config *ChainConfig, bc *BlockChain, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, totalUsedGas *big.Int) (*types.Receipt, evm.Logs, *big.Int, error) {
```
And now I feel like I'm just nitpicking... but exported vars and funcs should have comments, eg.
```
// ApplyMultiVmTransaction applies a transaction within SputnikVM...
func ApplyMultiVmTransaction(config ...
```

sorpaas commented:
@tzdybal Which one do you mean are the "incorrect" test cases?

tzdybal commented:
@sorpaas I mean the ones disabled in the test code long time ago.

sorpaas commented:
@tzdybal I created a new issue #427 for discussing those tests.

Everyone, please take a last look of the code here. If there's no objection, I'll merge it after 24 hours tomorrow.

rel (fix #416, part of milestone #372).

sorpaas commented:
Current build status:

* We use a build tag system. Current build command in README continues to work, but build without the sputnikvm library. Run the compiled binary with `--sputnikvm` argument gives out correct errors.
* Use this comment (https://github.com/ethereumproject/go-ethereum/pull/413#issuecomment-348792618) for build instructions for building with sputnikvm library.
* CI and deployment builds geth with sputnikvm library.

Current test status:

* Regression test against mainnet all passed.
* go-ethereum tests using classic vm is unaffected.
* go-ethereum tests using sputnikvm passed. 4 additional tests are disabled due to various reasons. Refer to the diff and commits for explanation.
